src/consensus.rs

```diff
@@ -448,144 +448,175 @@ impl Consensus {
-            t = Instant::now();
-            if d >= timeout {
-                timeout = Duration::from_millis(self.config.tick_period_ms);
-                // We drive Raft every `tick_period_ms`.
+            let updates = self.advance_node(previous_tick, tick_period)?;
+
+            let mut elapsed = previous_tick.elapsed();
+
+            while elapsed > tick_period {
                 self.node.tick();
-                // Try to reapply entries if some were not applied due to errors.
-                let store = self.node.store().clone();
-                let stop_consensus = store.apply_entries(&mut self.node)?;
+
+                previous_tick += tick_period;
+                elapsed -= tick_period;
+            }
+
+            if self.node.has_ready() {
+                let stop_consensus = self.on_ready()?;
+
                 if stop_consensus {
                     return Ok(());
                 }
-            } else {
-                timeout -= d;
-            }
-            let stop_consensus = self.on_ready()?;
-            if stop_consensus {
-                return Ok(());
+            } else if updates == Some(0) {
+                self.try_sync_local_state()?;
             }
         }
     }
 
-    fn try_sync_local_state(&mut self) -> anyhow::Result<()> {
-        if !self.node.has_ready() {
-            // No updates to process
-            let store = self.node.store();
-            let pending_operations = store.persistent.read().unapplied_entities_count();
-            if pending_operations == 0 && store.is_leader_established.check_ready() {
-                // If leader is established and there is nothing else to do on this iteration,
-                // then we can check if there are any un-synchronized local state left.
-                store.sync_local_state()?;
+    fn advance_node(
+        &mut self,
+        previous_tick: Instant,
+        tick_period: Duration,
+    ) -> anyhow::Result<Option<usize>> {
+        if previous_tick.elapsed() >= tick_period {
+            return Ok(None);
+        }
+
+        if self
+            .try_promote_learner()
+            .context("failed to promote learner")?
+        {
+            return Ok(Some(1));
+        }
+
+        let mut updates = 0;
+        let mut timeout_at = previous_tick + tick_period;
+
+        loop {
+            let message = match self.recv_update(timeout_at) {
+                Ok(message) => message,
+                Err(_) => break,
+            };
+
+            let is_conf_change = matches!(
+                message,
+                Message::FromClient(
+                    ConsensusOperations::AddPeer { .. } | ConsensusOperations::RemovePeer(_)
+                ),
+            );
+
+            if let Err(err) = self.advance_node_impl(message) {
+                log::warn!("{err}");
+                continue;
+            }
+
+            updates += 1;
+            timeout_at = Instant::now() + tick_period / 10;
```

Comment:
why `tick_period / 10`?

Reply:
TL;DR: No particular reason for it to be `tick_period / 10`, it just have to be a reasonably short interval relative to `tick_period`.

We kinda have two different situations with consensus:
- sometimes the nodes are lagging behind, so it gets a constant stream of updates from leader, and so we get a lot of messages in the `advance_node` queue
- but most of the time the nodes are up-to-date, and so updates are rather sparse
- when the node is lagging behind we want to push as many updates as fast as possible, so that's why we "batch" multiple `advance_node` calls
- but when updates are sparse (e.g., we only get pings and occasional consensus operation), we also don't want to wait 100ms before calling `on_ready`, because most operations in `raft` crate *does not* depend on `tick`, and if we do `advance_node`/`on_ready` in quick succession, consensus (on a whole cluster level) would progress *faster*

So what this whole `timeout_at` manipulation does is:
- on each iteration of an outer loop in consensus thread, we wait for the *first* event for up to `tick_period`
- but for each event *after* the first one, we only wait for `tick_period / 10`
- this way, when there are a lot of updates queued/incoming, we would pull a lot of them each iteration of outer loop
- but if when updates are sparse, we would wait for `tick_period / 10` for the next update and if there is none, we would go and call `on_ready`, so that the delay between `advance_node` and `on_ready` would be shorter when the load on consensus is low, so that consensus resolves faster

But overall, there is no particular reason for this timeout to be `tick_period / 10`. I used `Duration::from_millis(5)` initially. It could even be `Duration::ZERO` (with some additional tweaks). It just have to be something "relatively short compared to `tick_period`".